_posts/2020-09-30-ghelamco-alert.md

```diff
@@ -0,0 +1,824 @@
+This is defined in our serverless stack which handler should be invoked by the api gateway.  
+Following example illustrates how our lambda reads data from our dynamoDB table (using the environment variable from our config) and returns the result to our api gateway.  
+
+<div style="text-align: center;">
+  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/sls-lambda.PNG" width="auto" height="auto" target="_blank" class="image fit">
+</div>
+
+#### AWS API Gateway additional configuration
+{:.no_toc}
+
+We needed to add some additional configuration for our api gateway that was not possible to include in the functions part of our serverless template.  
+To make sure our application does not get CORS errors when it receives a 4XX or 5XX response from the api gateway we had set up CORS for these error responses.  
+As you can see below we added response headers to allow all origins and headers.  
+<div style="text-align: center;">
+  <img alt="Ghelamco-alert Cloudwatch Alarm" src="/img/2020-09-25-ghelamco-alert/sls-api.PNG" width="auto" height="auto" target="_blank" class="image fit">
+</div>
+
+### CICD pipeline Serverless
+
+Since the serverless framework is so easy to use, and we only had 1 environment available, we decided not to build a CICD pipeline for this part.  
+Instead, we just use the **serverless cli** to deploy and update our stack from our local development machines.  
+
+# Application flow
+To finish off the technical part of this blogpost I would like to show you the end-to-end workflow of the application.  
+For example let's say the user logs in to the web application and wants to **snooze an alert**.  
+
+1. We click snooze alert in the frontend web application
+2. This calls our api gateway endpoint **"/alert/snooze"**.
+3. This call is authorized by the **userpool JWT token** and **identitypool temporary AWS credential**.  
+4. This api gateway endpoint invokes an **AWS Lambda function** which creates a job and **job document in AWS iot** for the RPI.
+It also inserts a row into a second **dynamoDB table** (ghela-jobs) to register the job we created on the frontend.  
+5. The job gets send to the RPI and processed by our backend application, thus snoozing the alert in the **H2 database**.
+This will effectively turn the alert light off.    
+6. The RPI **synchronizes** his H2 database with the event **dynamoDB table**. Thus updating the status of the alert in the dynamoDB table.  
+7. This new status will then be reflected in the **frontend web application** so our user can see the result of his action.  
+
+This same workflow applies to all commands that can be sent from the frontend application to our RPI device.  
+This is a prime example of how a distributed system works and operates.  
+
+# Conclusion
+This internship gave me such a big cover of all best practices and new technologies, that this was a real eye-opening and educational experience.  
```

Comment:
```suggestion
This internship gave me such a big overview of all best practices and new technologies, that this was a real eye-opening and educational experience.  
```

Or: This internship covered such a large amount of best practices and new technologies that this was a real eye-opening educational experience.